### PR TITLE
[iOS] Fix data races in `RCTImageLoader` and `RCTNetworkTask` with shared atomic counters

### PR DESCRIPTION
**Pull request by hakonk**
## Summary:

In order to fix the data races described in https://github.com/facebook/react-native/issues/44715, I propose a simple solution by leveraging shared counter functions wherein `std::atomic` is the backing for the integer values.

## Changelog:

[iOS] [Fixed] - Implement shared atomic counters and replace static integers in `RCTImageLoader` and `RCTNetworkTask` that were accessed concurrently, which in some cases lead to data races.

## Test Plan:

Added unit tests for the counters in `RCTSharedCounterTests`.

### CODE REVIEW

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -863,6 +867,7 @@
 			isa = XCBuildConfiguration;
 			buildSettings = {
 				ALWAYS_SEARCH_USER_PATHS = NO;
+				CC = "";
```
Comment (cipolleschi) [inline packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj:870]: This change should be reverted.

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -893,6 +898,7 @@
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
 				COPY_PHASE_STRIP = NO;
+				CXX = "";
```
Comment (cipolleschi) [inline packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj:901]: This change should be reverted.

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -933,6 +939,8 @@
 					"${PODS_CONFIGURATION_BUILD_DIR}/React-graphics/React_graphics.framework/Headers",
 				);
 				IPHONEOS_DEPLOYMENT_TARGET = 13.4;
+				LD = "";
+				LDPLUSPLUS = "";
```
Comment (cipolleschi) [inline packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj:943]: This change should be reverted.

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -964,6 +972,7 @@
 			isa = XCBuildConfiguration;
 			buildSettings = {
 				ALWAYS_SEARCH_USER_PATHS = NO;
+				CC = "";
```
Comment (cipolleschi) [inline packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj:975]: This change should be reverted.

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -1027,6 +1037,8 @@
 					"${PODS_CONFIGURATION_BUILD_DIR}/React-graphics/React_graphics.framework/Headers",
 				);
 				IPHONEOS_DEPLOYMENT_TARGET = 13.4;
+				LD = "";
+				LDPLUSPLUS = "";
```
Comment (cipolleschi) [inline packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj:1041]: This change should be reverted.

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Great catch and smart solution!
I left some questions and a few minor changes requested.

**Comment by hakonk** [comment]
Hi @cipolleschi, thanks for the feedback! I will look into the comments and think of improvements. Will push changes when I have the time.

File: packages/react-native/React/Base/RCTSharedCounter.mm [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+#import "RCTSharedCounter.h"
+#import <atomic>
+#include <memory>
+
+NSUIntegerCounter RCTCreateAtomicNSUIntegerCounter(void) {
+  auto count = std::make_shared<std::atomic<NSUInteger>>(0);
```
[the remark this answers is not in the document]
Comment (hakonk) [reply packages/react-native/React/Base/RCTSharedCounter.mm]: My understanding is that this is correct. I have attempted to allow for testing this via [this commit](https://github.com/facebook/react-native/pull/45114/commits/880a33bf149a0da7ff6830d38a4f5b1809d09903)

File: packages/react-native/React/Base/RCTSharedCounter.h [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+#import <Foundation/Foundation.h>
+#import <React/RCTDefines.h>
+
+typedef NSUInteger (^NSUIntegerCounter)(void);
+typedef uint64_t (^UInt64Counter)(void);
+
+RCT_EXTERN NSUIntegerCounter RCTCreateAtomicNSUIntegerCounter(void);
```
[the remark this answers is not in the document]
Comment (hakonk) [reply packages/react-native/React/Base/RCTSharedCounter.h]: In the case of `RCTNetworkTask:24`, we create a static counter like so:
```objective-c++
static const auto generateRequestId = RCTCreateAtomicNSUIntegerCounter();
```
That is, the counter function is not static unless it is held statically. In turn, calling on `generateRequestId` will yield an incremented value. I believe it is more appropriate to name the variable that holds the block `generateRequestId`, as calling this block indeed increments the counter, while `RCTCreateAtomicNSUIntegerCounter` creates a new counter block where count is initialized to `0` for each call.

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj (hunk 5 above)
Comment (hakonk) [reply packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj]: [8c7547e](https://github.com/facebook/react-native/pull/45114/commits/8c7547ec8823356c1578e79384ec159b1e10c817)

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -994,6 +1003,7 @@
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
 				COPY_PHASE_STRIP = YES;
+				CXX = "";
```
[the remark this answers is not in the document]
Comment (hakonk) [reply packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj]: [8c7547e](https://github.com/facebook/react-native/pull/45114/commits/8c7547ec8823356c1578e79384ec159b1e10c817)

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj (hunk 4 above)
Comment (hakonk) [reply packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj]: [8c7547e](https://github.com/facebook/react-native/pull/45114/commits/8c7547ec8823356c1578e79384ec159b1e10c817)

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj (hunk 3 above)
Comment (hakonk) [reply packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj]: [8c7547e](https://github.com/facebook/react-native/pull/45114/commits/8c7547ec8823356c1578e79384ec159b1e10c817)

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj (hunk 2 above)
Comment (hakonk) [reply packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj]: [8c7547e](https://github.com/facebook/react-native/pull/45114/commits/8c7547ec8823356c1578e79384ec159b1e10c817)

File: packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj (hunk 1 above)
Comment (hakonk) [reply packages/rn-tester/RNTesterPods.xcodeproj/project.pbxproj]: [8c7547e](https://github.com/facebook/react-native/pull/45114/commits/8c7547ec8823356c1578e79384ec159b1e10c817)

**Comment by hakonk** [comment]
@cipolleschi please see the proposed changes above

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Thanks for verifying that the memory is deallocated.
For maintainability I'd rather remove the spy code, so the code is clearer and easier to understand and maintain.
Thanks for putting extra work in verifying the memory management! ❤️

**Comment by cipolleschi** [comment]
(also, remember to rebase on top of main, please! 🙏 )

**Comment by javache** [comment]
Since we already use Objective-C++ quite widely across the codebase, can we just replace this with std::atomic<int>?

**Comment by hakonk** [comment]
> Since we already use Objective-C++ quite widely across the codebase, can we just replace this with std::atomic<int>?

We certainly can. My approach was to generalize the solution, seeing it as one could make use of atomic counter functions for cases to come. I believe testing becomes easier as well, as one does not have to for example instantiate a class wherein a static count variable is used.

Is it more preferable from your point of view not to introduce the counters?

**Comment by hakonk** [comment]
> I'd rather rely on standard C++ fundamentals and use those directly, rather than using wrappers which involve shared_ptr, which have their own.
> 
> In pure Objective-C, we should also be able to use `OSAtomicIncrement64` as a much simpler solution.

@javache I have revised according to your review. Please see changes.

**Review by javache — APPROVED** [review]

**Pull request closed without merging**